### PR TITLE
fix(fs): mkdir("/") returns EINVAL instead of EEXIST

### PR DESCRIPTION
**Pull request by CharlieVinnie**
# fix: `mkdir("/")` returns `EINVAL` instead of `EEXIST`

## Summary

`busybox mkdir -p /tmp/foo` failed on riscv64 with "can't create directory '/': Invalid argument". The `-p` flag in busybox's `bb_make_directory` temporarily null-terminates the path at each `/` separator and calls `mkdir` on each intermediate component — the first call is always `mkdir("/", 0777)`. Linux returns `EEXIST` for this, which busybox tolerates and skips. StarryOS returned `EINVAL`, causing the entire `-p` operation to abort immediately.

## Root Cause

`FsContext::create_dir` calls `resolve_nonexistent`, which splits a path into `(parent_dir, filename)`. For paths with no filename component — `/`, `.`, `..` — `path.file_name()` returns `None`, so `resolve_nonexistent` returns `VfsError::InvalidInput` (EINVAL). This is the correct behaviour for `resolve_nonexistent` itself (it can't return a name to create), but `create_dir` propagated it directly to userspace instead of checking whether the path already exists.

## Fix

In `FsContext::create_dir` (`os/arceos/modules/axfs-ng/src/highlevel/fs.rs`), catch the `InvalidInput` case from `resolve_nonexistent` and resolve the path directly. If it resolves to an existing directory, return `AlreadyExists` (EEXIST). If it resolves to a non-directory node, return `NotADirectory` (ENOTDIR). This matches Linux kernel behaviour for `mkdirat` on root-like paths.

## Testing

Added three new test cases to `test-suit/starryos/normal/busybox/sh/busybox-tests.sh`:
- `busybox_mkdir`: `busybox mkdir -p /tmp/bb_mkd_one` — exercises the root-component path
- `busybox_mv`: `busybox mv` on a regular file — passes as a sanity check
- `busybox_rmdir`: `busybox rmdir` on a freshly created directory — depends on mkdir working

All three pass on riscv64.

## Architecture Notes

The bug manifested on riscv64 because the `busybox-bug` test was only wired up for riscv64. The fix is architecture-independent — `create_dir` is shared across all targets and the root-path edge case can be triggered on any arch.

### CODE REVIEW

**Comment by AsakuraMizu** [comment]
@Mivik Can you review this? Is this implemention good?

File: os/arceos/modules/axfs-ng/src/highlevel/fs.rs [hunk 1]
```diff
@@ -271,7 +271,23 @@ impl FsContext {
 
     /// Creates a new, empty directory at the provided path.
     pub fn create_dir(&self, path: impl AsRef<Path>, mode: NodePermission) -> VfsResult<Location> {
-        let (dir, name) = self.resolve_nonexistent(path.as_ref())?;
+        let path = path.as_ref();
+        let (dir, name) = match self.resolve_nonexistent(path) {
+            Ok(pair) => pair,
+            Err(VfsError::InvalidInput) => {
```
Comment (ZR233) [inline os/arceos/modules/axfs-ng/src/highlevel/fs.rs:281]: 这里把所有 `resolve_nonexistent` 的 `InvalidInput` 都当成“已存在目录”处理，会把空路径 `mkdir("")` 从应有的 `ENOENT` 变成 `EEXIST`。Linux/POSIX 下空 pathname 不是已存在目录，`mkdir("")` 返回 `ENOENT`；只有 `/`、`.`、`..` 这类能解析到既有目录的路径才应返回 `EEXIST`。建议先判断空路径并返回 `VfsError::NotFound`，再对可解析的无 filename 路径执行 `resolve(path)` 并映射为 `AlreadyExists`。

**Review by ZR233 — CHANGES_REQUESTED** [review]
发现 mkdir 路径边界条件的错误码回归，请按行内建议修复。

File: os/arceos/modules/axfs-ng/src/highlevel/fs.rs (hunk 1 above)
Comment (Charliewzy) [reply os/arceos/modules/axfs-ng/src/highlevel/fs.rs:281]: 谢谢，已修复，并添加了相应测例

File: test-suit/starryos/normal/bugfix/bug-mkdir-empty-path/qemu-riscv64.toml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+args = [
```
Comment (ZR233) [inline test-suit/starryos/normal/bugfix/bug-mkdir-empty-path/qemu-riscv64.toml]: 这个 `qemu-*.toml` 放在 `normal/bugfix/bug-mkdir-empty-path/` 下，但 xtask 的 normal 批量发现只扫描 `test-suit/starryos/normal` 的直接子目录；我本地运行 `cargo xtask starry test qemu --arch riscv64 -c bug-mkdir-empty-path` 也报 unknown case。按项目 Starry test-suit 约定，`bugfix` 已经是 grouped case，建议把 `/usr/bin/bug-mkdir-empty-path` 加到 `normal/bugfix/qemu-*.toml` 的 `test_commands`，或把该 case 移到 `normal/bug-mkdir-empty-path/` 作为直接 case。否则复现测例不会进入常规 Starry normal 批量覆盖。

File: test-suit/starryos/normal/bugfix/bug-mkdir-empty-path/qemu-riscv64.toml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    "virtio-net-pci,netdev=net0",
+    "-netdev",
+    "user,id=net0",
+]
+uefi = false
+to_bin = true
+shell_prefix = "root@starry:"
+shell_init_cmd = "/usr/bin/bug-mkdir-empty-path"
+success_regex = ['(?m)^TEST PASSED\s*$']
+fail_regex = ['(?i)\bpanic(?:ked)?\b', 'TEST \d FAILED']
```
Comment (ZR233) [inline test-suit/starryos/normal/bugfix/bug-mkdir-empty-path/qemu-riscv64.toml]: 这里的 `fail_regex` 是大小写敏感的 `TEST \d FAILED`，但 C 测例实际失败输出是 `Test 1 FAILED: ...` / `Test 2 FAILED: ...`。Starry QEMU runner 依赖 `fail_regex` 及时判定失败；当前写法不会匹配这些失败行，错误路径会变成等待 success 超时或诊断不清。建议改成例如 `(?m)^Test \d+ FAILED:` 或大小写不敏感的匹配，并同步其他架构的 qemu 配置。

**Review by ZR233 — CHANGES_REQUESTED** [review]
本地验证中，修复逻辑本身和复现程序可以覆盖 mkdir("") -> ENOENT、mkdir("/") -> EEXIST，但新增 StarryOS case 的布局不会被 normal 批量 suite 自动发现，失败匹配也抓不到测例实际输出。请先修正测试接入方式后再合入。

验证结果：
- cargo fmt --check: 通过。
- cargo xtask clippy --package ax-fs-ng: 通过，7/7 checks passed。
- cargo xtask starry test qemu --arch riscv64 -c bugfix/bug-mkdir-empty-path: 通过，输出包含 mkdir("") correctly returns ENOENT、mkdir("/") correctly returns EEXIST、TEST PASSED。
- cargo xtask starry test qemu --arch riscv64 -c bug-mkdir-empty-path: 失败，确认该新增 case 不是 direct normal case。

File: test-suit/starryos/normal/bugfix/bug-mkdir-empty-path/qemu-riscv64.toml (hunk 2 above)
Comment (Charliewzy) [reply test-suit/starryos/normal/bugfix/bug-mkdir-empty-path/qemu-riscv64.toml]: 谢谢，已修正

File: test-suit/starryos/normal/bugfix/bug-mkdir-empty-path/qemu-riscv64.toml (hunk 3 above)
Comment (Charliewzy) [reply test-suit/starryos/normal/bugfix/bug-mkdir-empty-path/qemu-riscv64.toml]: 谢谢，将测例并入 bugfix 批量测试中后不再需要自定义 fail_regex 了，已解决

**Review by ZR233 — APPROVED** [review]
## 审查结果：批准

修复 `mkdir` 对空路径和根路径的错误返回值，符合 POSIX 语义。

### 核心修复

1. `mkdir("")` → `ENOENT`（空路径返回 NotFound）
2. `mkdir("/")` 和 `mkdir(".")` → `EEXIST`（路径无文件名组件时，解析并检查已存在目录）
3. `mkdir("..")` → `EEXIST`（由 `resolve_nonexistent` 的 `AlreadyExists` 错误正确传播）

所有返回值与 Linux 行为一致。

### 测试覆盖

`bug-mkdir-empty-path` 测试用例覆盖了空路径、根路径和当前目录三种情况，errno 检查正确。

### 本地验证

- `cargo fmt --check`：通过
- `cargo xtask clippy --package starry-kernel`：失败来自 `ax-net-ng` 依赖，与 mkdir 变更无关
